add shape_type and base_material attributes to item definition

bell07 opened the pull request:
The information about the parentage of shaped nodes can be useful after the init stage. This PR adds the attributes shape_type and base_material to node/item definition to get this info available for other mods.
This change is primary for smart_inventory. If active the "not_in_creative_inventory" is ignored and the shaped items are available in a separate "shaped" group for easy access.
The change is generic enough for more use cases. I do not know the background of issue #70 but I think it is solvable by this change too.
Why I do not use the stairsplus.shapes_list? They are more mods that generates shaped things. See carpets or ts_doors as example. I prefer to implement a generic way for all instead of the "dependency hell".

sofar commented:
You're pushing maintenance overhead for a different mod down to other mods. 

I'm not convinced this is sincere, even. Dependecy "hell" only exists for your mod, and it's a trivially solvable problem, since you already know which mods would work with it.

so, thumbs up from me for this, sorry.

bell07 commented:
I do not see the maintenance overhead, if the patch is applied the stairsplus just provides additional information. Maybe if anyone defines a new shape type, there are 2 lines self-explainable code to be added.

This PR does not add any additional dependencies, and it does not break anything. It is just a small enhancement of item definition API to provide already existing information to other mods, like the not_in_creative_inventory group attribute (that is just provided but not used by moreblocks too.. and used by inventory mods).
smart_inventory does work fine without this change, the stairsplus shapes are ignored in this case since they are not_in_creative_inventory.

Ant there is need for this additional information. @VanessaE came before me with #70 and solved it as needed for them (without any discussion if it is useful or not). But the disadvantage in his change is: the change is for stairsplus only, my proposal is more general and applicable "as a standard" to different mods that creates shaped nodes. @VanessaE : How is your opinion?
I am open for proposals and ready to rewrite all my patches if the "shape_type" and "base_material" attributes can be obtained in a beter (unified) way.

> Dependecy "hell" only exists for your mod

I try to avoid the dependency hell by not implementing things which belong by design elsewhere. A feature in my mod uses the proposed enhancement. It the enhancement is not at place, the shaped items are just handled as "normal items". The dependency hell starts if I start to implement reading stairsplus.shapes_list, doors.registered_doors, start to create own lists for walls, stairs, slabs,....

Currently with all mods patched (minetest_game, stairsplus, carpets, ts_doors,...) I get 6623 shaped items in my inventory shaped group. If I activate the technic modpack in addition I get 7071 so I need to monitor all mods for stairsplus.register* usage and maintain the dependencies and items lists as a workaround if the stairsplus.register* does not provide the proposed (for me required) attributes. The same for walls,stairs, slabs,doors,... :-(

VanessaE commented:
What I had in mind for #70 was so that a mod could call upon Moreblocks/stairsplus for normal use, and also be able to override or remove the stairsplus objects associated with some specific node type in that mod.  Building_blocks (part of Homedecor modpack) does this with tar, since its tar node predates the one in Moreblocks.

It was not intended as a generic "shapes and materials" idea.

bell07 commented:
> It was not intended as a generic "shapes and materials" idea.

But your change does the same: "micro_", "panel_".. is the shape_type and k is the base_material. My proposal is to unify/standardize such attributes to be able to get this info independent from origin mod of the shaped item.

bell07 commented:
Now I got it. I seen in minetest_game there a group defined for each shape_type, so groups door, fence, stair, slab and wall already defined by minetest_game (why I did not seen it at first :-(). So I get the shape_type by selecting for a small list of groups for now. The idea about "base_material" I give up. It is a abstract simplicity that cannot match all use cases, the real "base_material" is the recipe if defined.
So I am fine with minetest_game for now. Thanks for your patience. I am still a newby.

But anyway in moreblocks/stairsplus the groups are not used/filled. Therefore I changed the PR implementation. micro, panel and slope does not exists in minetest_game but I added them too. I hope it is ok

sofar commented:
> I seen in minetest_game there a group defined for each shape_type, so groups door, fence, stair, slab and wall already defined by minetest_game 

No, they are defined for other reasons. E.g. fence and wall groups are defined because they are connected nodeboxes, and you need a group to make them connect.

Slab and stair groups exist for interaction purposes, as well as door.

These are not shape groups, they are internal code groups and should not be used for other purposes.

still :-1:

bell07 commented:
I give up. For now moreblocks shapes are just not supported by smart_inventory. If you have an Idea how I can detect the shape type for nodes generated by moreblocks, let me know. A long Items-list is still not an option.
